File: src/main/java/com/alphawallet/attestation/eip712/Eip712AttestationRequestWithUsage.java
package com.alphawallet.attestation.eip712;

import static com.alphawallet.attestation.eip712.Timestamp.DEFAULT_TIME_LIMIT_MS;
import static com.alphawallet.attestation.eip712.Timestamp.DEFAULT_TOKEN_TIME_LIMIT;

import com.alphawallet.attestation.AttestationRequestWithUsage;
import com.alphawallet.attestation.FullProofOfExponent;
import com.alphawallet.attestation.IdentifierAttestation.AttestationType;
import com.alphawallet.attestation.core.ExceptionUtil;
import com.alphawallet.attestation.core.SignatureUtility;
import com.alphawallet.attestation.core.URLUtility;
import com.alphawallet.attestation.core.Validateable;
import com.alphawallet.attestation.core.Verifiable;
import com.alphawallet.attestation.eip712.Eip712AttestationRequestWithUsageEncoder.AttestationRequestWUsageData;
import com.alphawallet.attestation.eip712.Eip712AttestationUsageEncoder.AttestationUsageData;
import java.io.IOException;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.bouncycastle.crypto.params.AsymmetricKeyParameter;
import org.tokenscript.eip712.Eip712Issuer;
import org.tokenscript.eip712.Eip712Validator;
import org.tokenscript.eip712.JsonEncodable;

public class Eip712AttestationRequestWithUsage extends Eip712Validator implements JsonEncodable,
    Verifiable, Validateable, TokenValidateable {
  private static final Logger logger = LogManager.getLogger(Eip712AttestationRequestWithUsage.class);

  private final long maxTokenValidityInMs;
  private final long acceptableTimeLimit;
  private final AttestationRequestWithUsage attestationRequestWithUsage;
  private final AttestationRequestWUsageData data;
  private final String jsonEncoding;
  private final AsymmetricKeyParameter userPublicKey;

  public Eip712AttestationRequestWithUsage(String attestorDomain, String identifier,
      AttestationRequestWithUsage attestationRequestWithUsage, AsymmetricKeyParameter signingKey) {
    this(attestorDomain, DEFAULT_TIME_LIMIT_MS, DEFAULT_TOKEN_TIME_LIMIT,
        identifier, attestationRequestWithUsage, signingKey);
  }

  public Eip712AttestationRequestWithUsage(String attestorDomain,
      long acceptableTimeLimit, long maxTokenValidityInMs, String identifier,
      AttestationRequestWithUsage attestationRequestWithUsage, AsymmetricKeyParameter signingKey) {
    super(attestorDomain, new Eip712AttestationRequestWithUsageEncoder());
    try {
      this.acceptableTimeLimit = acceptableTimeLimit;
      this.maxTokenValidityInMs = maxTokenValidityInMs;
      this.attestationRequestWithUsage = attestationRequestWithUsage;
      this.jsonEncoding = makeToken(identifier, attestationRequestWithUsage, signingKey);
      this.userPublicKey = retrieveUserPublicKey(jsonEncoding, AttestationRequestWUsageData.class);
      this.data = retrieveUnderlyingObject(jsonEncoding, AttestationRequestWUsageData.class);
    } catch (Exception e ) {
      throw ExceptionUtil.throwException(logger,
          new IllegalArgumentException("Could not encode object"));
    }
    constructorCheck();
  }

  public Eip712AttestationRequestWithUsage(String attestorDomain, String jsonEncoding) {
    this(attestorDomain, DEFAULT_TIME_LIMIT_MS, DEFAULT_TOKEN_TIME_LIMIT, jsonEncoding);
  }

  public Eip712AttestationRequestWithUsage(String attestorDomain,
      long acceptableTimeLimit, long maxTokenValidityInMs, String jsonEncoding) {
    super(attestorDomain, new Eip712AttestationRequestWithUsageEncoder());
    try {
      this.acceptableTimeLimit = acceptableTimeLimit;
      this.maxTokenValidityInMs = maxTokenValidityInMs;
      this.jsonEncoding = jsonEncoding;
      this.userPublicKey = retrieveUserPublicKey(jsonEncoding, AttestationRequestWUsageData.class);
      this.data = retrieveUnderlyingObject(jsonEncoding, AttestationRequestWUsageData.class);
      this.attestationRequestWithUsage = new AttestationRequestWithUsage(URLUtility.decodeData(data.getPayload()));
    } catch (Exception e ) {
      throw ExceptionUtil.throwException(logger,
          new IllegalArgumentException("Could not decode object"));
    }
    constructorCheck();
  }

  void constructorCheck() throws IllegalArgumentException {
    if (!verify()) {
      throw ExceptionUtil.throwException(logger,
          new IllegalArgumentException("Could not verify Eip712 use attestation"));
    }
  }

  String makeToken(String identifier, AttestationRequestWithUsage attestationRequestWithUsage,
      AsymmetricKeyParameter signingKey) throws IOException {
    Eip712Issuer issuer = new Eip712Issuer<AttestationUsageData>(signingKey, encoder);
    String encodedUseAttestation = URLUtility.encodeData(attestationRequestWithUsage.getDerEncoding());
    Timestamp now = new Timestamp();
    Timestamp expirationTime = new Timestamp(now.getTime() + maxTokenValidityInMs);
    AttestationRequestWUsageData data = new AttestationRequestWUsageData(
        encoder.getUsageValue(), identifier, encodedUseAttestation, now, expirationTime);
    return issuer.buildSignedTokenFromJsonObject(data, domain);
  }

  public String getIdentifier() {
    return data.getIdentifier();
  }

  public AsymmetricKeyParameter getUserPublicKey() {
    return userPublicKey;
  }

  public FullProofOfExponent getPok() {
    return attestationRequestWithUsage.getPok();
  }

  public AttestationType getType() {
    return attestationRequestWithUsage.getType();
  }

  public AsymmetricKeyParameter getSessionPublicKey() {
    return attestationRequestWithUsage.getSessionPublicKey();
  }

  @Override
  public String getJsonEncoding() {
    return jsonEncoding;
  }

  /**
   * Verify that an attestation can be issued. I.e. the nonce is not expired
   */
  @Override
  public boolean checkValidity() {
    if (!testNonceAndDescription(acceptableTimeLimit)) {
      logger.error("The object can no longer be used as attestation request. Nonce validation failed");
      return false;
    }
    return true;
  }

  /**
   * Verify that the object can be used as a usage token. I.e. the token timestamp has not expired.
   * Note that the object can still be used as a token after the nonce for issuance has expired.
   */
  @Override
  public boolean checkTokenValidity() {
    Timestamp time = new Timestamp(data.getTimestamp());
    time.setValidity(maxTokenValidityInMs);
    if (!time.validateAgainstExpiration(Timestamp.stringTimestampToLong(data.getExpirationTime()))) {
      logger.error("The object can no longer be used as a request token. It is expired.");
      return false;
    }
    // Nonce validation must still happen since this also verifies user's address and receiver's domain
    if (!testNonceAndDescription(maxTokenValidityInMs)) {
      logger.error("The object can no longer be used as a request token. Nonce validation failed");
      return false;
    }
    return true;
  }

  private boolean testNonceAndDescription(long timeLimit) {
    long nonceMinTime = Timestamp.stringTimestampToLong(data.getTimestamp()) - timeLimit;
    long nonceMaxTime = Timestamp.stringTimestampToLong(data.getTimestamp()) + timeLimit;
    if (!Nonce.validateNonce(attestationRequestWithUsage.getPok().getNonce(),
        SignatureUtility.addressFromKey(userPublicKey), domain, new Timestamp(nonceMinTime), new Timestamp(nonceMaxTime))) {
      logger.error("Nonce validation failed");
      return false;
    }
    if (!data.getDescription().equals(encoder.getUsageValue())) {
      logger.error("Description field is incorrect");
      return false;
    }
    return true;
  }

  @Override
  public boolean verify() {
    if (!attestationRequestWithUsage.verify()) {
      logger.error("Could not verify signature");
      return false;
    }
    return true;
  }
}

File: src/main/javascript/crypto/src/libs/Eip712AttestationRequestWithUsage.ts
import {Eip712Validator} from "./Eip712Validator";
import {JsonEncodable} from "../intefaces/JsonEncodable";
import {Verifiable} from "./Verifiable";
import {Validateable} from "./Validateable";
import {Eip712AttestationUsage} from "./Eip712AttestationUsage";
import {TokenValidateable} from "./TokenValidateable";
import {Eip712AttestationRequest} from "./Eip712AttestationRequest";
import { AttestationRequestWithUsage } from "./AttestationRequestWithUsage";
import {KeyPair} from "./KeyPair";
import {base64ToUint8array, hexStringToBase64Url} from "./utils";
import {FullProofOfExponent} from "./FullProofOfExponent";
import {SignatureUtility} from "./SignatureUtility";
import {Nonce} from "./Nonce";
import {AttestationRequest} from "./AttestationRequest";
import {Eip712Token} from "./Eip712Token";
import {Timestamp} from "./Timestamp";

export class Eip712AttestationRequestWithUsage extends Eip712Token implements JsonEncodable,
    Verifiable, Validateable, TokenValidateable {
    // public static DEFAULT_TOKEN_TIME_LIMIT: number = Eip712AttestationUsage.DEFAULT_TOKEN_TIME_LIMIT;
    // public static DEFAULT_TIME_LIMIT_MS: number = Eip712AttestationRequest.DEFAULT_TIME_LIMIT_MS;

    private Eip712UserDataTypes: {name: string, type: string}[]  = [
        {name: 'payload', type: 'string'},
        {name: 'description', type: 'string'},
        {name: 'identifier', type: 'string'},
        {name: 'timestamp', type: 'string'},
        {name: 'expirationTime', type: 'string'},
    ]

    protected data: {
        payload: string,
        description: string,
        identifier: string,
        timestamp: string,
        expirationTime: string,
    }

    private Eip712UserDataPrimaryName: string = "AttestationRequestWUsage";
    private Eip712UserDataDescription: string = "Prove that the \"identity\" is the identity hidden in attestation contained in\"payload\""
        + " and use this to authorize usage of local, temporary keys.";

    private maxTokenValidityInMs: number;
    private acceptableTimeLimit: number;
    private attestationRequestWithUsage:AttestationRequestWithUsage;

    private jsonEncoding: string;
    private userPublicKey: KeyPair;
    private userKey: KeyPair;

    constructor(userKey: KeyPair = null, acceptableTimeLimit:number = Timestamp.DEFAULT_TIME_LIMIT_MS, maxTokenValidityInMs:number = Timestamp.DEFAULT_TOKEN_TIME_LIMIT) {
        super();
        this.userKey = userKey;

        this.acceptableTimeLimit = acceptableTimeLimit;
        this.maxTokenValidityInMs = maxTokenValidityInMs;
    }


    public async fromData(attestorDomain:string, acceptableTimeLimit:number = Timestamp.DEFAULT_TIME_LIMIT_MS, maxTokenValidityInMs:number = Timestamp.DEFAULT_TOKEN_TIME_LIMIT  ,identifier:string,
    attestationRequestWithUsage: AttestationRequestWithUsage, signingKey: KeyPair = null) {

        this.setDomain(attestorDomain);
        if (signingKey) {
            this.userKey = signingKey;
        }
        try {
            this.acceptableTimeLimit = acceptableTimeLimit;
            this.maxTokenValidityInMs = maxTokenValidityInMs;
            this.attestationRequestWithUsage = attestationRequestWithUsage;
            this.jsonEncoding = await this.makeToken(identifier, attestationRequestWithUsage);
        } catch ( e ) {
            console.log(e);
            throw new Error("Could not encode object");
        }

        try {
            this.fillJsonData(this.jsonEncoding);
        } catch ( e ) {
            throw new Error("Could not decode object");
        }
    }

    public Eip712AttestationRequestWithUsage(attestorDomain:string,
        acceptableTimeLimit:number, maxTokenValidityInMs: number, jsonEncoding:string) {
            //TODO
            // super(attestorDomain);
        try {
            this.acceptableTimeLimit = acceptableTimeLimit;
            this.maxTokenValidityInMs = maxTokenValidityInMs;
            this.jsonEncoding = jsonEncoding;

            this.fillJsonData(this.jsonEncoding);

        } catch ( e ) {
            console.log(e);
            throw new Error("Could not decode object");
        }
    }

    fillJsonData(json: string){
        if (!json) throw new Error('Empty json');

        this.jsonEncoding = json;
        let tokenData = JSON.parse(json);
        let signatureInHex = tokenData.signatureInHex;

        let jsonSigned = JSON.parse(tokenData.jsonSigned);
        this.eip712DomainData = jsonSigned.domain;
        this.data = jsonSigned.message;

        try {
            let publicKey = SignatureUtility.recoverPublicKeyFromTypedMessageSignature(jsonSigned, signatureInHex);
            this.userPublicKey = KeyPair.fromPublicHex(publicKey.substr(2));
            // console.log('Eip712 withUsage restored address: ' + this.userPublicKey.getAddress());
        } catch (e){
            let m = "Recover Address failed with error:" + e;
            console.log(m)
            throw new Error(m);
        }

        if (!this.attestationRequestWithUsage){
            this.attestationRequestWithUsage = AttestationRequestWithUsage.fromBytes(base64ToUint8array( this.data.payload));
        }

        this.constructorCheck();
    }

    constructorCheck() {
        if (!this.verify()) {
            throw new Error("Could not verify Eip712 use attestation");
        }
    }

    async makeToken(identifier: string, attestationRequestWithUsage: AttestationRequestWithUsage) {

        if (!this.userKey) {
            await SignatureUtility.connectMetamaskAndGetAddress();
        }

        let ts = new Timestamp().getTimeAsString();
        let expirationTime = new Timestamp(Date.now() + this.maxTokenValidityInMs).getTimeAsString();

        let userData = {
            payload: hexStringToBase64Url(attestationRequestWithUsage.getDerEncoding()),
            description: this.Eip712UserDataDescription,
            timestamp: ts,
            identifier: identifier,
            expirationTime: expirationTime,
        };


        return await SignatureUtility.signEIP712WithBrowserWallet(this.domain, userData, this.Eip712UserDataTypes, this.Eip712UserDataPrimaryName, this.userKey );
    }

    public getIdentifier(): string {
        return this.data.identifier;
    }

    public getUserPublicKey(): KeyPair {
        return this.userPublicKey;
    }

    public getPok(): FullProofOfExponent {
        return this.attestationRequestWithUsage.getPok();
    }

    public getType(): number {
        return this.attestationRequestWithUsage.getType();
    }

    public getSessionPublicKey(): KeyPair {
        return this.attestationRequestWithUsage.getSessionPublicKey();
    }

    public getJsonEncoding(): string {
        return this.jsonEncoding;
    }

    /**
     * Verify that an attestation can be issued. I.e. the nonce is not expired
     */
    public checkValidity(): boolean {
        if (!this.testNonceAndDescription(this.acceptableTimeLimit)) {
            return false;
        }
        return true;
    }

    /**
     * Verify that the object can be used as a usage token. I.e. the token timestamp has not expired.
     * Note that the object can still be used as a token after the nonce for issuance has expired.
     */
    public checkTokenValidity():boolean {
        let time:Timestamp = new Timestamp(this.data.timestamp);
        time.setValidity(this.maxTokenValidityInMs);
        if (!time.validateAgainstExpiration(Timestamp.stringTimestampToLong(this.data.expirationTime))) {
            console.log('time.validateAgainstExpiration filed');
            return false;
        }
        // Nonce validation must still happen since this also verifies user's address and receiver's domain
        if (!this.testNonceAndDescription(this.maxTokenValidityInMs)) {
            return false;
        }
        return true;
    }

    private testNonceAndDescription(timeLimit: number): boolean {
        if (!timeLimit) {
            throw new Error('timeLimit required');
        }
        let nonceMinTime: number  = Timestamp.stringTimestampToLong(this.data.timestamp) - timeLimit;
        let nonceMaxTime: number = Timestamp.stringTimestampToLong(this.data.timestamp) + timeLimit;
        if (!new Nonce().validateNonce(
            this.attestationRequestWithUsage.getPok().getNonce(),
            this.userPublicKey.getAddress(),
            this.domain,
            nonceMinTime,
            nonceMaxTime
        )) {
            return false;
        }
        if (this.data.description !== this.Eip712UserDataDescription ) {
            return false;
        }
        return true;
    }

    public verify(): boolean {
        if (!this.attestationRequestWithUsage.verify()) {
            return false;
        }
        return true;
    }
}
